Mathlib/NumberTheory/NumberField/CanonicalEmbedding/NormLeOne.lean

```diff
@@ -0,0 +1,165 @@
+
+Mainly, we prove that this is bounded, its frontier has volume zero and compute its volume.
+
+## Strategy of proof
+
+The proof is loosely based on the strategy given in [D. Marcus, *Number Fields*][marcus1977number].
+
+1. since `NormLeOne K` is norm-stable, in the sense that
+  `normLeOne K = normAtAllPlaces⁻¹' (normAtAllPlaces '' (normLeOne K))`,
+  see `normLeOne_eq_primeage_image`, it's enough to study the subset
+  `normAtAllPlaces '' (normLeOne K)` of `realSpace K`.
+
+2. A description of `normAtAllPlaces '' (normLeOne K)` is given by `normAtAllPlaces_normLeOne`, it
+  is the set of `x : realSpace K`, nonnegative at all places, whose norm is nonzero and `≤ 1` and
+  such that `logMap x` is in the `fundamentalDomain` of `basisUnitLattice K`.
+  Note that, here and elsewhere, we identify `x` with its image in `mixedSpace K` given
+  by `mixedSpaceOfRealSpace x`.
+
+3. In order to describe the inverse image in `realSpace K` of the `fundamentalDomain` of
+  `basisUnitLattice K`, we define the map `expMap : realSpace K → realSpace K` that is, in
+  some way, the right inverse of `logMap`, see `logMap_expMap`.
+
+-/
+
+variable (K : Type*) [Field K]
+
+open NumberField NumberField.InfinitePlace NumberField.mixedEmbedding NumberField.Units
+
+namespace NumberField.mixedEmbedding
+
+section normAtAllPlaces
+
+variable [NumberField K]
+
+variable {K}
+
+theorem logMap_normAtAllPlaces (x : mixedSpace K) :
+    logMap (mixedSpaceOfRealSpace (normAtAllPlaces x)) = logMap x :=
+  logMap_eq_of_normAtPlace_eq
+    fun w ↦ by rw [normAtPlace_mixedSpaceOfRealSpace (normAtPlace_nonneg w x)]
+
+theorem norm_normAtAllPlaces (x : mixedSpace K) :
+    mixedEmbedding.norm (mixedSpaceOfRealSpace (normAtAllPlaces x)) = mixedEmbedding.norm x := by
+  simp_rw [mixedEmbedding.norm_apply,
+    normAtPlace_mixedSpaceOfRealSpace (normAtAllPlaces_nonneg _ _)]
+
+theorem normAtAllPlaces_mem_fundamentalCone_iff {x : mixedSpace K} :
+    mixedSpaceOfRealSpace (normAtAllPlaces x) ∈ fundamentalCone K ↔ x ∈ fundamentalCone K := by
+  simp_rw [fundamentalCone, Set.mem_diff, Set.mem_preimage, logMap_normAtAllPlaces,
+    Set.mem_setOf_eq, norm_normAtAllPlaces]
+
+end normAtAllPlaces
+
+section normLeOne_def
+
+variable [NumberField K]
+
+/--
+The set of elements of the `fundamentalCone` of `norm ≤ 1`.
+-/
+abbrev normLeOne : Set (mixedSpace K) :=
+  {x | x ∈ fundamentalCone K ∧ mixedEmbedding.norm x ≤ 1}
+
+variable {K} in
+theorem mem_normLeOne (x : mixedSpace K) :
+    x ∈ normLeOne K ↔ x ∈ fundamentalCone K ∧ mixedEmbedding.norm x ≤ 1 := Set.mem_sep_iff
+
+theorem normLeOne_eq_primeage_image :
+    normLeOne K = normAtAllPlaces⁻¹' (normAtAllPlaces '' (normLeOne K)) := by
+  refine subset_antisymm (Set.subset_preimage_image _ _) ?_
+  rintro x ⟨y, hy₁, hy₂⟩
+  rw [mem_normLeOne, ← normAtAllPlaces_mem_fundamentalCone_iff, ← norm_normAtAllPlaces,
+    ← mem_normLeOne] at hy₁ ⊢
+  rwa [← hy₂]
+
+open scoped Classical in
+theorem normAtAllPlaces_normLeOne :
```

Comment:
Can you write this using intersection of sets? I suspect that the API for `A ∩ B` is more developed than  those for `{x | f x ∧ g x}` (of course those are 100% the same, but let's avoid defeq abuse as much as possible).